connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java

```diff
@@ -26,16 +28,26 @@
 public class TransformationChain<R extends ConnectRecord<R>> {
 
     private final List<Transformation<R>> transformations;
+    private final RetryWithToleranceOperator retryWithToleranceOperator;
 
     public TransformationChain(List<Transformation<R>> transformations) {
+        this(transformations, RetryWithToleranceOperator.NOOP_OPERATOR);
+    }
+
+    public TransformationChain(List<Transformation<R>> transformations, RetryWithToleranceOperator retryWithToleranceOperator) {
         this.transformations = transformations;
+        this.retryWithToleranceOperator = retryWithToleranceOperator;
     }
 
     public R apply(R record) {
         if (transformations.isEmpty()) return record;
 
-        for (Transformation<R> transformation : transformations) {
-            record = transformation.apply(record);
+        for (final Transformation<R> transformation : transformations) {
+            final R current = record;
+
+            // execute the operation
+            record = retryWithToleranceOperator.execute(() -> transformation.apply(current), Stage.TRANSFORMATION, transformation.getClass());
```

Comment:
This would be more readable if the operator parameter is last, since that's the one parameter that's likely to be a lambda.

Reply:
responded below.